### ISSUE TITLE
Comparison Guides of other Frameworks

### ISSUE THREAD

**Issue by LadyBluenotes** [solidjs/solid-docs-next#496]

**Comment by atilafassina** [comment]
That's a steep slippery slope 😅

SolidStart and:
- [ ] Next.js
- [ ] Remix
- [ ] Analog
- [ ] SvelteKit
- [ ] Nuxt

**Comment by birkskyum** [comment]
.. and of course Astro, and maybe solid-router+vite without Start (people _often_ ask if they'd be fine like this in discord) :)

**Comment by amirhhashemi** [comment]
I'm very skeptical about this idea. I often see comparisons like this in libraries that want to set themselves apart from similar ones or showcase unique features that others don't have. But honestly, I don't think SolidStart would come out looking great in such comparisons. We have fewer features compared to more mature and established frameworks.

That's not to say we should hide SolidStart’s shortcomings. However, creating a comparison like this tends to reduce a framework to a mere list of features, which could be misleading. Many of SolidStart's features are also present in other frameworks, but with nuanced differences in their implementation.

I also don't think this is an effective approach to assist programmers in deciding whether to adopt SolidStart. Generally, the choice of a framework is influenced more by the ecosystem and familiarity with the underlying UI library (Solid) than by a list of features. Most framework features can be implemented or replicated by users if they're missing. If a framework lacks a feature that some may consider essential, it's likely that the programmer has already done thorough research beyond what we can offer on this page.

The effort required to create such a comprehensive comparison is another issue. We're talking about 5+ large, constantly evolving frameworks. Writing an unbiased comparison that covers everything and keeping it updated is a big task. That's a lot of work for something that probably won't help much. Even if the community helped, I don't think it'd be worth the effort.

The only legitimate use case I see for a comparison is if, let's say, someone coming from Next.js wants to know whether we provide image optimization. In such cases, we have a great community that can provide answers. By focusing our efforts on improving our current docs, finding information like this shouldn't be difficult.

**Comment by atilafassina** [comment]
@amirhhashemi 

I'm often one to say I prefer teaching Solid from first-principles. But I think SolidStart has a proper a few reasons to have a section such as this.

> But honestly, I don't think SolidStart would come out looking great in such comparisons.

The fact that you, despite having more than enough experience with SolidStart, feels this way about it is reason why I think we should have something like "SolidStart Purpose", explaining why the lean core and how to fill the "lack" of functionality and how. So, to counter your point, no, I don't agree that SolidStart would come in looking bad in these comparisons.

> The effort required to create such a comprehensive comparison is another issue. We're talking about 5+ large, constantly evolving frameworks. Writing an unbiased comparison that covers everything and keeping it updated is a big task. That's a lot of work for something that probably won't help much. Even if the community helped, I don't think it'd be worth the effort.

I'm 100% with you here, I think we should have a few comparisons in the direction of a migration guide/helper from popular inbound frameworks to begin with (Next.js, React Router). Specially because we have some weird acronyms and practices that change meaning/labels between frameworks ("on-request middleware", "global middleware", "route middleware", "proxy", ISR, SSR, etc) - different names for similar things are source 1 of confusion for new users.

And finally another entry to answer the new most common question "What's the difference between Tanstack Start w/Solid and SolidStart?"

**Comment by amirhhashemi** [comment]
@atilafassina 

When it comes to comparisons, we are expected to be fair and thorough. We can't only focus on the positive aspects; we must also address the negatives. We must mention that SolidStart lacks some of the nice features of Next.js and has a much smaller ecosystem. In this context, lack of features is not a feature, it's a weakness. That's why I think SolidStart would look bad in such comparisons, and for the wrong reasons.

It's just suicide to compare SolidStart to frameworks like Next.js, Nuxt.js, and SvelteKit. SolidStart is newer and smaller. Of course, it's going to look worse. If we want SolidStart to gain attraction, that's not the way to go.

I think the best way to approach this is to improve our marketing. All the best aspects of SolidStart should be on our home page. And given the unique design of SolidStart, I completely agree with you that we should explain design decisions throughout our docs.

When someone in Discord asks such a broad question as "what's the difference between Next.js and SolidStart" I automatically categorieze that person into one of three categories: 1. Newbie 2. Lazy 3. Both. It's easy to answer this question in Discord. We focus on positive stuff and encourage that person to read our docs:

1. SolidStart is based on Solid, which is a smaller and more performant UI library.
2. It's router-agnostic, so you can use other routers if you don't like `solid-router`.
3. It has very good data-fetching utilities, so you might not need Tanstack Query.
4. It supports all the basic stuff that you need, and you can deploy it everywhere.

But we can't put that in our documentation. Questions like these are better addressed in Discord.

I also disagree with creating migration guides for frameworks that come from completely different backgrounds. Migrating Next.js to SolidStart? Where do we even begin? The two are fundamentally different. Instead, we should focus on improving our own documentation. For example, if middleware works differently in SolidStart, we should explain it better. If a specific feature of another framework can be replicated, we should create a how-to guide for it.

We might consider Tanstack Start + Solid as an exception and add a migration guide and comparison for it if they share significant similarities. But I'm not familiar with Tanstack Start.

**Comment by atilafassina** [comment]
@amirhhashemi 

> When it comes to comparisons, we are expected to be fair and thorough. 

I think that's a fundamental problem in our docs. We want them to be perfect, but then we don't do because it's hard or take time. What's left is that we have nothing.  Continuous improvement is better than delayed perfection.

> We can't only focus on the positive aspects; we must also address the negatives. We must mention that SolidStart lacks some of the nice features of Next.js and has a much smaller ecosystem. In this context, lack of features is not a feature, it's a weakness. That's why I think SolidStart would look bad in such comparisons, and for the wrong reasons.

I never said to focus on positive aspects only. Perhaps there are some assumptions going on here with things I left unsaid. Of course I know that monolithic frameworks go above and beyond what Start offers - I'm saying that's not always a good thing.

> It's just suicide to compare SolidStart to frameworks like Next.js, Nuxt.js, and SvelteKit. SolidStart is newer and smaller. Of course, it's going to look worse. If we want SolidStart to gain attraction, that's not the way to go.

Disagree with the suicide part. And, reading reddit, discord, etc. Lack of onboarding and documented APIs is generally the n.1 reason people don't try or only use for side-projects. Again, because we focus on delayed perfection.

> I think the best way to approach this is to improve our marketing. 
Docs are n.1 marketing tool for a project such as Solid. That's where people go to learn or to be guided. It's most important connection between framework/library devs and their users - and right now, we're leaving **a lot** unsaid.

> All the best aspects of SolidStart should be on our home page. 

A very early decision in the docs was to push information from the landing page to docs. The landing page is just an entrypoint. Consider it an "elevator pitch". Moreover, people don't go to the landing page to learn how the framework works, they go to the docs.

> When someone in Discord asks such a broad question as "what's the difference between Next.js and SolidStart" I automatically categorieze that person into one of three categories: 1. Newbie 2. Lazy 3. Both. It's easy to answer this question in Discord. We focus on positive stuff and encourage that person to read our docs:
> 
> 1. SolidStart is based on Solid, which is a smaller and more performant UI library.
> 2. It's router-agnostic, so you can use other routers if you don't like `solid-router`.
> 3. It has very good data-fetching utilities, so you might not need Tanstack Query.
> 4. It supports all the basic stuff that you need, and you can deploy it everywhere.

Hm. Again, I never said we should address such broad question. That perhaps is a function of our upcoming MCP server or something that we could (and should!) bring back Orama's AI function to the docs.

> I also disagree with creating migration guides for frameworks that come from completely different backgrounds. Migrating Next.js to SolidStart? Where do we even begin? The two are fundamentally different. 

We start by common use-cases. They have significant overlap in user intentions. Again, delayed perfection.

> Instead, we should focus on improving our own documentation. For example, if middleware works differently in SolidStart, we should explain it better. 

Yes, most of our docs entries should teach things from first principles. But we need to accept most of Solid users are immigrants from other ecosystems, our job (the docs!) is to make the adoption smoother.

> If a specific feature of another framework can be replicated, we should create a how-to guide for it.

Exactly what I meant from the beginning. 

> We might consider Tanstack Start + Solid as an exception and add a migration guide and comparison for it if they share significant similarities. But I'm not familiar with Tanstack Start.

That I fundamentally disagree with. It sounds lazy to me, we don't pick the docs we write by what's easier to do, we pick them by what our users need most. 

1. Tantsack Start surfaces architecture questions: we need to make our docs more thorough on that (first principles) and _perhaps_ add something specific.
2. Most of our users show up with pre-existing Next.js knowledge, we need to guide them how to do things in the Start way, how to accomplish what they know how to do in Next.js - this is a basic pedagogic principle: you teach focused on your student's background.

**Comment by LadyBluenotes** [comment]
@atilafassina @amirhhashemi 

To insert myself into this conversation...

I agree with Amir in that I'm reluctant to have a guide on migration from other frameworks. This would require a lot of upkeep and effort from the existing docs team and considering there's few people who contribute to them, I think our time is better spent on the actual framework rather than comparing what we do or don't provide compared to other features.

I think, what might be a more appropriate response for this, is that we focus on documenting the features of SolidStart better. Maybe once we get to a point where we have all the features / guides implemented that give a better idea of the capabilities of the framework we can consider an idea like this (or if a community member wishes to volunteer their time into writing a guide, we can work with them from there).

This feels like a goal for people who have a larger list of contributors for their docs. TanStack, Astro, Vue, Svelte, etc. all have that and thus are more likely to be able to sustain things like that. We can't even manage to get the basics of the framework (and core) documented without a ton of effort because we have such a small team and we're all pulled in a bunch of different directions. 

So while a migration guide can be "nice to have", we simply do not have the manpower to be able to execute it. I think we'd be doing ourselves a disservice in focusing on how someone might migrate from another to this, when we don't even have fully fleshed out docs as it is. Much to Atila's point, we need to get our docs to be "good enough" (not perfect) before we can even bring this further into the conversation.

**Comment by amirhhashemi** [comment]
@atilafassina 

> I think that's a fundamental problem in our docs. We want them to be perfect, but then we don't do because it's hard or take time. What's left is that we have nothing. Continuous improvement is better than delayed perfection.

I'm not saying it should be perfect at all. But it should at least meet some basic standards. In this case, writing a fair and comprehensive comparison is the minimum expectation. Otherwise, we may mislead people into choosing the wrong framework, which I think is far worse than having nothing at all.

Also, given that it's a big commitment to keep a page updated with 5+ constantly changing frameworks, I think if we are not sure that we can completely maintain it ourselves, and just hope that someone will come and improve it, we are going to end up with an outdated guide that misleads people and just adds another burden to the team. 

> I never said to focus on positive aspects only. Perhaps there are some assumptions going on here with things I left unsaid. Of course I know that monolithic frameworks go above and beyond what Start offers - I'm saying that's not always a good thing.

I don't understand why SolidStart being lean is perceived as a positive thing by most users. Can you provide some examples?

> Disagree with the suicide part. And, reading reddit, discord, etc. Lack of onboarding and documented APIs is generally the n.1 reason people don't try or only use for side-projects. Again, because we focus on delayed perfection.

> Docs are n.1 marketing tool for a project such as Solid. That's where people go to learn or to be guided. It's most important connection between framework/library devs and their users - and right now, we're leaving a lot unsaid.

I initially didn't consider SolidStart because it had bad documentation. That's the main reason I'm contributing to improve it. But I don't sure how a comparison page is going to help. If anything, it increases the workload for the docs team and lead to a decline in the overall quality of the core docs.

> A very early decision in the docs was to push information from the landing page to docs. The landing page is just an entrypoint. Consider it an "elevator pitch". Moreover, people don't go to the landing page to learn how the framework works, they go to the docs.

I agree. My point is that the key advantages of SolidStart should be marketed better. If we need to create a comparison page to highlight what sets SolidStart apart, that indicates a marketing/documentation issue. However, a comparison page may not be the best place to communicate this information.

> We start by common use-cases. They have significant overlap in user intentions. Again, delayed perfection.

There are many common use cases across different frameworks, and each framework has multiple documentation pages focused on teaching these basic features. This is not a small project.

Why not encourage our users to read the main documentation and learn the framework instead of creating this page?

> Yes, most of our docs entries should teach things from first principles. But we need to accept most of Solid users are immigrants from other ecosystems, our job (the docs!) is to make the adoption smoother.

Improving the documentation *does* smooth the adoption. We shouldn't treat programmers as if they don't know how to read normal documentation. That's the most effective way to learn SolidStart.

> Exactly what I meant from the beginning.

I think my proposal is fundamentally different. In my suggestion, these guides would not be specific to any framework and should be included in our core documentation rather than on a separate page.

**Comment by amirhhashemi** [comment]
@atilafassina @LadyBluenotes 

I was reviewing old issues to see what I could do when I came across this one. The reason I’m restarting the conversation, despite its inactivity for a while, is that I fundamentally disagree with this idea. I don’t this it’s even a nice-to-have.

I've shared my reasons, but the actual goal and target audience for this page are still unclear to me. Maybe if we establish that first, we can have a more productive conversation.

**Comment by atilafassina** [comment]
I think there's a misunderstanding here. I'm not advocating on the urgency of this Guide, just that it's important to have for **the most common one(s)** (not five, not multiple... one or two).

I agree that a comprehensive documentation on fundamentals and features of the framework is more important and would be most desirable. Perhaps the criticism I started here about delayed perfection is misplaced and meant for another moment, so I'll stop talking about it now.

> We shouldn't treat programmers as if they don't know how to read normal documentation. That's the most effective way to learn SolidStart.

Never meant to say people don't know how to read docs ( "normal"  or not), but different people learn differently and have different intents. We can't one-size-fits-all teaching. Allowing them to easily find out what they're looking for is the main task for a docs app/site

that means catering for different use-cases (the one learning from scratch in a greenfield environment, the one trying to get something done, the one refactoring a brownfield app, etc.)

---

So, while I'm **not** saying this is a top priority, in the context we have now: this is a necessity - not because I want it or even because others have, but because most questions we get on discord and reddit are from people with previous knowledge that are confused and trying to shoehorn patterns that don't really fit (auth being the biggest example that comes to mind, but hardly the only one)

is this the most critical shortcoming of the docs? No, I agree with you.
But it is something the docs should address.

**Comment by atilafassina** [comment]
Sorry, I sent my last comment before reading this one, @amirhhashemi 

> [@atilafassina](https://github.com/atilafassina) [@LadyBluenotes](https://github.com/LadyBluenotes)
> 
> I was reviewing old issues to see what I could do when I came across this one. The reason I’m restarting the conversation, despite its inactivity for a while, is that I fundamentally disagree with this idea. I don’t this it’s even a nice-to-have.
> 
> I've shared my reasons, but the actual goal and target audience for this page are still unclear to me. Maybe if we establish that first, we can have a more productive conversation.

I agree that this conversation has dragged for too long. This is not a priority and neither something either of us shuld be working now.

Perhaps, at some point when you have implemented more of your ideas, a migration guide becomes obsolete. But as things are now, there's confusion to mitigate and addressing it head-on is a shortcut / "temporary fix"
